packages/language-server/test/completion-items-utils.ts

```diff
@@ -0,0 +1,234 @@
+import { forEach } from "lodash";
+import { expect } from "chai";
+import {
+  TextDocument,
+  TextDocumentPositionParams,
+  Position,
+  TextDocumentIdentifier,
+  CompletionItem,
+  TextEdit,
+  Range
+} from "vscode-languageserver";
+import { UI5SemanticModel } from "@ui5-language-assistant/semantic-model-types";
+import { expectExists } from "@ui5-language-assistant/test-utils";
+import { getCompletionItems } from "../src/completion-items";
+
+/** Return the first part of a tag name suggestion insert text */
+export function getTagName(textEdit: TextEdit | undefined): string | undefined {
+  if (textEdit === undefined) {
+    return undefined;
+  }
+  const result = /^([^> ]*)/.exec(textEdit.newText);
+  return result?.[1];
+}
+
+/** Use ⇶ to mark the cursor position */
+export function getSuggestions(
+  xmlSnippet: string,
+  ui5SemanticModel: UI5SemanticModel
+): CompletionItem[] {
+  const { document, position } = getXmlSnippetDocument(xmlSnippet);
+  const uri: TextDocumentIdentifier = { uri: "uri" };
+  const textDocPositionParams: TextDocumentPositionParams = {
+    textDocument: uri,
+    position: position
+  };
+
+  const suggestions = getCompletionItems(
+    ui5SemanticModel,
+    textDocPositionParams,
+    document
+  );
+  // Check that all returned suggestions will be displayed to the user
+  assertSuggestionsAreValid(suggestions, xmlSnippet);
+  return suggestions;
+}
+
+function createTextDocument(languageId: string, content: string): TextDocument {
+  return TextDocument.create("uri", languageId, 0, content);
+}
+
+function getXmlSnippetDocument(
+  xmlSnippet: string
+): { document: TextDocument; position: Position } {
+  const xmlSnippetWithoutRanges = xmlSnippet
+    .replace(/⭲/g, "")
+    .replace(/⭰/g, "");
+  const xmlText = xmlSnippetWithoutRanges.replace("⇶", "");
+  const offset = xmlSnippetWithoutRanges.indexOf("⇶");
+  const document: TextDocument = createTextDocument("xml", xmlText);
+  const position: Position = document.positionAt(offset);
+  return { document, position };
+}
+
+/** Use ⭲ to mark range start and ⭰ to mark range end */
+export function getRanges(xmlSnippet: string): Range[] {
+  const RANGE_START_CHAR = "⭲";
+  const RANGE_END_CHAR = "⭰";
+
+  const { document } = getXmlSnippetDocument(xmlSnippet);
+  const ranges: Range[] = [];
+  let xmlTextWithRanges = xmlSnippet.replace("⇶", "");
+  while (xmlTextWithRanges.indexOf(RANGE_START_CHAR) !== -1) {
```

Comment:
I would have attempted to implement this with a **none greedy** and **sticky** regExp
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky

- > https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Quantifiers (look for "greedy" word)

But it is clear enough so no need to change